Add retry mechanism to CosmosDB Database creation

msexxeta opened the pull request:
## Description

This is a workaround to the emulator container advertising it is ready when in fact it is not. 

Fixes #9326 
Fixes #7178


## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [x] No **fixes existing unit test** https://github.com/radical/aspire/blob/d57970b465f75fbe805bf101125ee9e4ef69882c/tests/Aspire.Hosting.Azure.Tests/AzureCosmosDBEmulatorFunctionalTests.cs#L271
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [x] No

msexxeta committed:
Add retry mechanism to Database creation

eerhardt commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -108,21 +108,34 @@ private static IResourceBuilder<AzureCosmosDBResource> RunAsEmulator(this IResou
-
-            foreach (var database in builder.Resource.Databases)
+            var databasesCreated = false;
+            while (!databasesCreated && !ct.IsCancellationRequested)
```
Would it make sense to use Polly here?

msexxeta replied:
I changed it to use polly with 5 seconds of retries and added a health check for database and container creation

msexxeta committed:
use polly for database creation retry and add health check for database and container creation

msexxeta committed:
Merge branch 'main' into patch-2

davidfowl commented:
Discussing with the cosmos team adding a health endpoint inside of the emulator to make this easier.

msexxeta committed:
Update AzureCosmosDBExtensions.cs

Try fixing failing test after merging main

msexxeta committed:
Merge branch 'dotnet:main' into patch-2

msexxeta committed:
Merge branch 'main' into patch-2

msexxeta committed:
increase retry timeout for cosmos db

davidfowl commented:
@eerhardt @sebastienros can you review this for 9.4 plz

msexxeta commented:
I'm don't understand why `Aspire.Hosting.Azure.Tests.AzureCosmosDBEmulatorFunctionalTests.VerifyWaitForOnCosmosDBEmulatorBlocksDependentResources(usePreview: False)` is failing, is it possible that `.OnResourceReady` can be triggered before `.OnConnectionStringAvailable`? Or that if there are no references used, will `.OnConnectionStringAvailable` be skipped?

eerhardt commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -106,28 +112,45 @@ private static IResourceBuilder<AzureCosmosDBResource> RunAsEmulator(this IResou
+                        await db.CreateContainerIfNotExistsAsync(containerProperties, cancellationToken: ct).ConfigureAwait(false);
+                    }
                 }
+                creationState = HealthCheckResult.Healthy();
```
This isn't really how health checks are supposed to work. They should be checking that the resource can be connected to correctly. This is saying once the db/containers are all created, assume it is healthy. But there may be other connection issues.

eerhardt committed:
Enable disabled test

msexxeta commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -153,11 +176,37 @@ static CosmosClient CreateCosmosClient(string connectionString)
                 {
                     clientOptions.ConnectionMode = ConnectionMode.Gateway;
                     clientOptions.LimitToEndpoint = true;
+                    clientOptions.CustomHandlers.Add(new CosmosClientRetryHandler());
```
Should have the same result, only difference is that this client is also used for health checks, so the health checks would also retry with the custom handler in here.

msexxeta replied:
The workaround I initially made for my project only retried the `ReadAccountAsync()` until it worked, then the subsequent db and container creation calls also worked without retries, so in theory that would suffice. I wanted to make sure that the databases will be created, though, so they are also retried, and if they're never created, the database should not become healthy.

msexxeta committed:
use inline retry pipeline

sebastienros committed:
Remove unnecessary ReadAccount call

msexxeta committed:
Merge branch 'dotnet:main' into patch-2

msexxeta committed:
update cosmosdb tests

msexxeta committed:
increase retry timeouts

msexxeta commented:
can someone continue the test run so I know if these changes fixed the issues? @sebastienros @davidfowl

msexxeta committed:
Update run-tests.yml

increase hang dump timeout for cosmos container startup

eerhardt commented on .github/workflows/run-tests.yml:
```diff
@@ -16,11 +16,11 @@ on:
         required: false
         type: string
-        default: "15m"
+        default: "21m"
```
If it is that slow, then let's keep the test disabled. I'd rather keep our CI loop fast than waiting on a flakey emulator.

msexxeta replied:
reverted and disabled that test

msexxeta commented:
I guess disable the tests again (at least the ones that create the emulator twice in a row) but keep the workaround so that it at least works for most people, because right now it is just simply broken.

msexxeta commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -106,28 +112,58 @@ private static IResourceBuilder<AzureCosmosDBResource> RunAsEmulator(this IResou
+            
+                                foreach (var container in database.Containers)
+                                {
+                                    var containerProperties = container.ContainerProperties ?? new ContainerProperties
+                                    {
+                                        Id = container.ContainerName,
+                                        PartitionKeyPaths = container.PartitionKeyPaths
+                                    };
```
I had to add that back in because removing it caused errors (that probably no one noticed since the test was already failing before this workaround).
Maybe it works without it now, but if `container.ContainerProperties` is properly set then the fallback doesn't run anways.

msexxeta committed:
revert extending test run durations

msexxeta committed:
Disable flaky test

msexxeta committed:
disable all cosmos emulator tests

radical commented on tests/Aspire.Hosting.Azure.Tests/AzureCosmosDBEmulatorFunctionalTests.cs:
```diff
@@ -128,15 +128,15 @@ await pipeline.ExecuteAsync(async token =>
     }
 
-    [Theory(Skip = "Using CosmosDB emulator in integration tests leads to flaky tests - https://github.com/dotnet/aspire/issues/5820")]
+    [Theory(Skip = "CosmosDB Emulator takes too long to spin up, spinning up twice consecutively will almost always exceed the global test timeout")]
```
Use the `[OuterloopTest(..)]` instead of skipping it. That will be then run by `.github/workflows/specialized-test-runner.yml` .

msexxeta committed:
move cosmosdb emulator tests to outer loop

eerhardt commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -106,28 +112,58 @@ private static IResourceBuilder<AzureCosmosDBResource> RunAsEmulator(this IResou
+                sp => cosmosClient ?? throw new InvalidOperationException("CosmosClient is not initialized."),
+                name: healthCheckKey
+            )
+            .AddCheck(creationHealthCheckKey, () => creationState);
```
This doesn't work because it creates a deadlock. `OnResourceReady` doesn't fire until the resource is "healthy". This check is making it being "healthy" dependent on `OnResourceReady` firing.

<img width="2723" height="345" alt="Image" src="https://github.com/user-attachments/assets/6e1d5409-9f24-4b1b-93cb-14df184cf636" />

msexxeta replied:
I see. In that case, would it suffice to move the database creation code up into the `OnConnectionStringAvailable` callback?

msexxeta replied:
that would be a major refactoring compared to how it is done right now and not my intention with this PR, I don't know enough about the internals of aspire to do that. This is a fix since this stuff is broken.

msexxeta replied:
As far as I saw it, the databases just inherit the health status from their parent resource. I did not intend to change that here, just fix stuff.

msexxeta committed:
move container creation to OnConnectionStringAvailable

Increased the maximum retry attempts for creating databases and containers from 1200 to 3600. Removed unnecessary initialization of container properties.

msexxeta committed:
revert test to use TaskCompletionSource

msexxeta committed:
revert to using TaskCompletionSource

Meir017 commented:
> Discussing with the cosmos team adding a health endpoint inside of the emulator to make this easier.

@davidfowl any updates on this?
